test/api_container_list_test.go

```diff
@@ -25,33 +30,152 @@ func (suite *APIContainerListSuite) SetUpTest(c *check.C) {
 	PullImage(c, busyboxImage125)
 }
 
-// TestListOk test api is ok with default parameters.
-func (suite *APIContainerListSuite) TestListOk(c *check.C) {
-	cname := "TestListOk"
+// TestListAll test api with -all parameters, empty filter.
+func (suite *APIContainerListSuite) TestListAll(c *check.C) {
+	containerA := "TestListAllContainerA"
+	resA := command.PouchRun("run", "-d", "--name", containerA, busyboxImage125, "top").Assert(c, icmd.Success)
+	defer DelContainerForceMultyTime(c, containerA)
+	containerAID := strings.TrimSpace(resA.Combined())
 
-	resp, err := CreateBusybox125Container(cname, "top")
-	c.Assert(err, check.IsNil)
-	CheckRespStatus(c, resp, 201)
+	containerB := "TestListAllContainerB"
+	command.PouchRun("create", "--name", containerB, busyboxImage125, "top").Assert(c, icmd.Success)
+	defer DelContainerForceMultyTime(c, containerB)
 
-	defer DelContainerForceMultyTime(c, cname)
+	// all: false
+	success, got, _ := getContainerListOK(c, "", false)
+	c.Assert(success, check.Equals, true)
+	c.Assert(len(got), check.Equals, 1)
+	// check fields
+	c.Assert(got[0].ID, check.Equals, containerAID)
+	c.Assert(got[0].Image, check.Equals, busyboxImage125)
+	c.Assert(got[0].ImageID, check.Equals, busyboxImage125ID)
+	c.Assert(len(got[0].Names), check.Equals, 1)
+	c.Assert(got[0].Names[0], check.Equals, containerA)
 
-	// NOTE: need to list all here because the container is not running
-	q := url.Values{}
-	q.Set("all", "true")
-	resp, err = request.Get("/containers/json", request.WithQuery(q))
+	// all: true
+	success, got, _ = getContainerListOK(c, "", true)
+	c.Assert(success, check.Equals, true)
+	c.Assert(len(got), check.Equals, 2)
+}
+
+// TestListFilterInvalid test invalid filter.
+func (suite *APIContainerListSuite) TestListFilterInvalid(c *check.C) {
+	success, _, errResp := getContainerListOK(c, "foo", false)
+	c.Assert(success, check.Equals, false)
+	err := util.PartialEqual(errResp.Message, "invalid character")
 	c.Assert(err, check.IsNil)
-	CheckRespStatus(c, resp, 200)
 
-	var got []types.Container
-	err = request.DecodeBody(&got, resp.Body)
+	success, _, errResp = getContainerListOK(c, "{\"foo\":[\"bar\"]}", false)
```

Comment:
This is for API test. When test this in cli, we use key=value.
While in API test, the filter data has been transferred to json.

Just comment to record this.